File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

/**

 */

class Time{
  int start, end;
  public Time(int start, int end){
    this.start = start;
    this.end = end;
  }

}

class Pair{
  int time;
  int value;

  public Pair(int time, int value) {
    this.time = time;
    this.value = value;
  }

  @Override
  public String toString() {
    return "Pair{" +
        "time=" + time +
        ", value=" + value +
        '}';
  }
}
class Solution {
  public int solution(String[][] book_time) {
    int answer = 0;
    List<Time> times = new ArrayList<>();

    for (String[] singleTime : book_time){
      String[] startTime = singleTime[0].split(":");
      String[] endTime = singleTime[1].split(":");
      int startTimeInt = (Integer.parseInt(startTime[0]) * 60) + Integer.parseInt(startTime[1]);
      int endTimeInt = (Integer.parseInt(endTime[0]) * 60) + Integer.parseInt(endTime[1]) + 10;

      times.add(new Time(startTimeInt, endTimeInt));
    }

    int cnt = 0;

    List<Pair> list = new ArrayList<>();

    for(Time t : times){
      list.add(new Pair(t.start, 1));
      list.add(new Pair(t.end, -1));
    }

    Collections.sort(list, (a, b) ->{
      if(a.time == b.time) return a.value - b.value;
      return a.time - b.time;
    });

    //System.out.println(list);
    for(Pair p : list){
      cnt += p.value;
      answer = Math.max(answer, cnt);
    }

    return answer;
  }
}

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.js
function solution(book_time) {
  var answer = 0;
  var formattedTime = [];

  function format(time){
    let [startHour, startMin] = time[0].split(":");
    let [endHour, endMin] = time[1].split(":");

    formattedTime.push([parseInt(startHour,10)*60 + parseInt(startMin,10) , parseInt(endHour, 10) * 60 + parseInt(endMin, 10) + 10]);
  }

  for(let b of book_time){
    format(b);
  }

  //formattedTime.sort((a,b) => a[0] - b[0]);
  
  let map = new Map();
  for(let [start, end] of formattedTime){
    map.set(start, (map.get(start) || 0) + 1);
    map.set(end, (map.get(end) || 0) - 1);
  }

  let sortedMap = new Map([...map.entries()].sort((a,b) => a[0] - b[0]));
  

  let tmp = 0;

  for(let [key , value] of sortedMap){
    //console.log(key, value, tmp);
    tmp += value;
    answer = Math.max(tmp, answer);
  }
  
  //console.log(answer)
  return answer;
}
